python/tests/test_tc.py

```diff
@@ -0,0 +1,434 @@
+    # Performance is not expected to be as good as when using 2 kernels.
+    #
+    def test_group_norm_fused(self):
+        group_normalization = """
+            def group_normalization(
+                float(N, G, D, H, W) I, float(G, D) gamma, float(G, D) beta) -> (Sum, SumSq, O)
+            {
+                Sum(n, g) +=! I(n, g, r_d, r_h, r_w)
+              SumSq(n, g) +=! I(n, g, r_d, r_h, r_w) * I(n, g, r_d, r_h, r_w)
+                O(n, g, d, h, w) = gamma(g, d)
+                    * ( I(n, g, d, h, w) - Sum(n, g) / (D * H * W))
+                    * rsqrt( (SumSq(n, g) / (D * H * W) - Sum(n, g) * Sum(n, g)) + 1e-5 )
+                    + beta(g, d)
+            }
+        """
+
+        N, G, D, H, W = 32, 32, 4, 56, 56
+        T = tc.define(
+            group_normalization,
+            TC.generate_options_from_tuning(tuner_config=tuner_config))
+        I, gamma, beta = (
+            torch.randn(N, G, D, H, W, device='cuda'),
+            torch.randn(G, D, device='cuda'),
+            torch.randn(G, D, device='cuda'))
+        Sum, SumSq, O = T.group_normalization(I, gamma, beta)
+        # TODO: test results!!!
+
+    #
+    # This test implements group normalization as 2 TC kernels
+    # (Note: loop collapsing modulo conforming strides is an easy TC-level
+    #  transformation with nice expected benefits).
+    #
+    def test_group_norm_2kernels(self):
+        group_normalization = """
+            def moments(float(N, K) I) -> (mean, var) {
+                # var = E(x^2) - mean^2.
+                mean(n) +=! I(n, r_k)
+                 var(n) +=! I(n, r_k) * I(n, r_k)
+                mean(n)  = mean(n) / (K)
+                 var(n)  =  var(n) / (K) - mean(n) * mean(n)
+            }
+
+            def group_normalization(
+                float(N, G, D, H, W) I, float(G, D) gamma, float(G, D) beta,
+                float(N, G) mean, float(N, G) var) -> (O)
+            {
+                O(n, g, d, h, w) = gamma(g, d)
+                    * ( I(n, g, d, h, w) - mean(n, g) )
+                    * rsqrt( var(n, g) + 1e-5 )
+                    + beta(g, d)
+            }
+        """
+
+        N, G, D, H, W = 32, 32, 4, 56, 56
+        I, gamma, beta = (
+            torch.randn(N, G, D, H, W, device='cuda'),
+            torch.randn(G, D, device='cuda'),
+            torch.randn(G, D, device='cuda'))
+
+        T = tc.define(
+            group_normalization,
+            TC.generate_options_from_tuning(tuner_config=tuner_config))
+
+        # First occurrence triggers tuning
+        mean, var = T.moments(I.view((N * G, -1)))
+        out, = T.group_normalization(
+            I, gamma, beta, mean.view((N, G)), var.view((N, G)))
+
+        # Subsequent occurrences do not
+        # But we can still choose to deactivate tuning and use naive options if
+        # ever the need arises.
+        T.generate_implicit_mapping_options = TC.generate_naive_options
```

Comment:
I don't like this part.   It rummages in some state internal to `T` in an inconsistent way (`.define` uses the result of `TC.generate_naive_options()` _call_, which is presumably a function as well; here, you just assign `generate_naive_options` to something without calling it; other examples use `TC.generate_options_from_tuning(...)` that _requires_ parameters which you cannot pass here because you are not calling the function).

Reply:
it's unnecessary I'll kill